### ISSUE TITLE
Block witness generation / verification and stateless execution

### ISSUE THREAD

**Issue by web3-developer** [status-im/nimbus-eth1#2327]
Prior work in Nimbus implemented both generation and verification of block witnesses based on the specs here: https://github.com/ethereum/portal-network-specs/blob/01a49a8c9bf08121ecde1b9270a6f2f679cb2568/witness.md

A backup of the stateless code is in this branch for future reference: https://github.com/status-im/nimbus-eth1/tree/stateless-backup. At some point we may want to re-enable these features.

The AccountCache / AccountLedger holds a data structure called the WitnessCache which records every touched account and storage slot for each block. After executing a block of transactions the code supports getting the keys and building a block witness from these keys. 

The idea is that these generated witnesses would be shared by block producers and used by stateless clients in order to verify blocks without having to store the state locally.

**Comment by web3-developer** [comment]
Here is the PR which removed this code for the time being: https://github.com/status-im/nimbus-eth1/pull/2295

**Comment by bhartnett** [comment]
> Still going over the design space, but its likely that the EL will not need to do the networking/client work like ress does and instead it would be done on the CL side -- here is a rough proof of concept: [sigp/lighthouse#7755](https://github.com/sigp/lighthouse/pull/7755)

Yes I agree that the Ress RLPx sub-protocol won't be needed/used in protocol for Ethereum but we may end up implementing it or something similar just to get our proof of concept fully working. This project is just the first step before we update it to implement binary tries and then later the zk-stateless work.

Makes sense that the CL could do the proof verification without an EL however something to keep in mind is that it will be valuable for the EL to have access to the proofs and state diffs so that we can implement light RPC full nodes which can provide trustless access to the state without having to depend on a third party RPC provider. Vitalik mentioned the idea of extending a stateless EL client to store the full (or partial) state without storing the merkle branches which should provider a x2 storage reduction (see [here](https://ethresear.ch/t/a-local-node-favoring-delta-to-the-scaling-roadmap/22368#p-54385-medium-term-priority-stateless-verification-3)). In order for the stateless EL client to follow the chain and update its copy of the state it should verify these proofs and then update its state using the proven state diffs.
